t/apicast-policy-upstream.t

```diff
@@ -0,0 +1,246 @@
+    content_by_lua_block {
+      local expected = "service_token=token-value&service_id=42&usage%5Bhits%5D=2&user_key=uk"
+      local args = ngx.var.args
+      if args == expected then
+        ngx.exit(200)
+      else
+        ngx.log(ngx.ERR, expected, ' did not match: ', args)
+        ngx.exit(403)
+      end
+    }
+  }
+--- configuration
+{
+  "services": [
+    {
+      "id": 42,
+      "backend_version":  1,
+      "backend_authentication_type": "service_token",
+      "backend_authentication_value": "token-value",
+      "proxy": {
+        "api_backend": "http://example:80/",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "GET", "metric_system_name": "hits", "delta": 2 }
+        ],
+        "policy_chain": [
+          {
+            "name": "apicast.policy.upstream",
+            "configuration":
+              {
+                "rules": [
+                  { "regex": "/i_dont_match", "url": "http://example.com" },
+                  { "regex": "/", "url": "http://test:$TEST_NGINX_SERVER_PORT" },
+                  { "regex": "/abc", "url": "http://example.com" }
+                ]
+              }
+          },
+          { "name": "apicast.policy.apicast" }
+        ]
+      }
+    }
+  ]
+}
+--- upstream
+  location /a_path {
+     content_by_lua_block {
+       ngx.say('yay, api backend');
+     }
+  }
+--- request
+GET /a_path?user_key=uk
+--- response_body
+yay, api backend
+--- error_code: 200
+--- no_error_log
+[error]
+
+=== TEST 3: non-matching rules
+In this example, none of the rules match, so the request will go to the
+upstream in 'api_backend'.
+--- backend
+  location /transactions/authrep.xml {
+    content_by_lua_block {
+      local expected = "service_token=token-value&service_id=42&usage%5Bhits%5D=2&user_key=uk"
+      local args = ngx.var.args
+      if args == expected then
+        ngx.exit(200)
+      else
+        ngx.log(ngx.ERR, expected, ' did not match: ', args)
+        ngx.exit(403)
+      end
+    }
+  }
+--- configuration
+{
+  "services": [
+    {
+      "id": 42,
+      "backend_version":  1,
+      "backend_authentication_type": "service_token",
+      "backend_authentication_value": "token-value",
+      "proxy": {
+        "api_backend": "http://test:$TEST_NGINX_SERVER_PORT",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "GET", "metric_system_name": "hits", "delta": 2 }
+        ],
+        "policy_chain": [
+          {
+            "name": "apicast.policy.upstream",
+            "configuration":
+              {
+                "rules": [
+                  { "regex": "/i_dont_match", "url": "http://example.com" },
+                  { "regex": "/i_dont_either", "url": "http://example.com" },
+                  { "regex": "/nope", "url": "http://example.com" }
+                ]
+              }
+          },
+          { "name": "apicast.policy.apicast" }
+        ]
+      }
+    }
+  ]
+}
+--- upstream
+  location /a_path {
```

Comment:
This can match `/a_path/whatever`. Would be good to at least define it as `location = /a_path`. But I'd also go the extra step and print `ngx.var.uri` in the response. So we can for example verify the args.

Reply:
I added a luassert test that checks the value of `ngx.var.request`. It turns out that args are not sent when the url of the rule contains a path.

Reply:
👍 I knew this was going to be a bit tricky to send everything correctly.

Reply:
I solved this, but it looks like I have a similar problem with POST bodies.